Prerender workers

Dantemss opened the pull request:
For: https://github.com/openstax/unified/issues/1797

Instructions:
1. Deploy the infrastructure stack in https://github.com/openstax/unified-deployment/pull/239 (already done in the sandbox)
2. `cd rex-web`
3. Export the following environment variables (you can save this as .env.local for example, then source it):
```
export REACT_APP_CODE_VERSION="test"
export REACT_APP_RELEASE_ID="test/2022-01-12"
export PUBLIC_URL="/rex/releases/$REACT_APP_RELEASE_ID"
export REACT_APP_ENV="production"
export PRERENDER_IMAGE_TAG=$(tr / - <<< $REACT_APP_RELEASE_ID)
```
4. Build and tag an image for the prerender target: `docker build . --target prerender -t 373045849756.dkr.ecr.us-east-2.amazonaws.com/rex:$PRERENDER_IMAGE_TAG`
5. Login to ECR: `aws ecr get-login-password --region us-east-2 | docker login --username AWS --password-stdin 373045849756.dkr.ecr.us-east-2.amazonaws.com`
6. To assume the Builder and Manager roles, you'd need to modify the openstax-users account, but you can use `sandbox:full-admin` instead.
7. As the Builder role (or full-admin), push the image: `docker push 373045849756.dkr.ecr.us-east-2.amazonaws.com/rex:$PRERENDER_IMAGE_TAG`
8. As the Manager role (or full-admin), run: `yarn prerender:fleet`

TomWoodward committed:
spike lambda renderer

TomWoodward committed:
working on node errors

TomWoodward committed:
i think this is in a working state now

Dantemss committed:
Added prerender manager script and copied the lambda script as the worker script

Dantemss committed:
Added worker code; Handle the sitemap as well

Dantemss committed:
Removed lambda-specific stuff

Dantemss committed:
Restored the original prerender script

Dantemss committed:
Commented out sitemap code and restored code that waits 1 minute and checks that the work queue message attributes are 0

Dantemss committed:
Fixed lint and typecheck errors; Actually delete processed messages from the queue

Dantemss committed:
Reverted change to script/entry.js

Dantemss committed:
Fixed some mistakes in the manage script; Added more logging

Dantemss committed:
Repurposed Tom's lambda Dockerfile
Fixed a bunch of errors in the prerendering script

Dantemss committed:
Fixed crash when trying to get messages from an empty queue

Dantemss committed:
Removed local proxy; Fixed lint errors

Dantemss committed:
Made the saveS3Page function async

Dantemss committed:
Reduced S3 concurrency to prevent errors and fixed Cluster deletion

Dantemss committed:
Use c6i instaces instead of t2

Dantemss committed:
Reverted credentials code

Dantemss committed:
Use node work_threads in the work script so a single instance of the script can use multiple cores and avoid network errors due to CPU contention

Dantemss committed:
Fixed errors in the threading code

Dantemss committed:
Send a single page to each thread rather than sets of 10

Dantemss committed:
Since we wait for the threads on each loop iteration to send the delete messages, there will always be threads available when we get 10 new messages, so simplify the logic

Dantemss committed:
Use asyncPool for prepareBooks
Use a promise for the work queues rather than polling

Dantemss committed:
Handle different message types
Queue and work sitemap jobs

Dantemss committed:
Write sitemaps to S3 instead of locally

Dantemss committed:
Upload sitemaps with text/xml content type

Dantemss committed:
Removed stats that don't work with distributed jobs

Dantemss committed:
Don't block the queuing thread waiting for all work threads to complete each batch of messages

Dantemss committed:
Removed more unused stats

Dantemss committed:
Create a separate concourse dir for prerender-workers

Dantemss committed:
Exclude already-uploaded files from verify-production-urls

Dantemss committed:
Replaced echo with print to fix shellcheck issue

Dantemss committed:
Fixed lint error that is only caught on CI for some reason

Dantemss commented:
Not entirely sure where the `src/app/navigation` changes are from... maybe I need to rebase something?
Also this changes one thing in the actual build if merged which is `verify-production-urls`. It should work, but I haven't tested it with Concourse so we'd have to test.

Dantemss committed:
Improved some comments

Dantemss commented:
The navigation changes were actually in the lambda PR... if they don't belong let me know and I can take them out.

TomWoodward commented on script/prerender-workers/work.ts:
```diff
@@ -0,0 +1,262 @@
+  }
+}
+
+async function sqsHeartbeat(messages: Message[]) {
```
what is sqs heartbeat?

Dantemss replied:
I should probably write this as a comment in there :)

TomWoodward commented on script/server/index.ts:
```diff
@@ -6,6 +6,8 @@ import serveStatic from 'serve-static';
 import config from '../../src/config';
 import setupProxy from '../../src/setupProxy';
 
+export { Server } from 'http';
```
whats this for?

Dantemss replied:
Probably a leftover from when I was using the proxy. I'll take it out.

Dantemss commented on Dockerfile:
```diff
@@ -26,11 +26,25 @@ RUN curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | b
+COPY . /code
+RUN yarn build
+
+ENTRYPOINT ["yarn"]
```
My intention was for the ENTRYPOINT and CMD to apply only to the prerender target, which is not used in the CI, but you are right that I have not tested if that is true. I'll see if I can test that this is the case.

Dantemss replied:
I have tested and confirmed that ENTRYPOINT and CMD apply only to the prerender target:

- When building an image with `docker build .` then running `docker run image-id` I get no message and exit code 0.
- When building an image with `docker build .` then running `docker run image-id echo Hello` I get `Hello`.
- When building an image with `docker build . --target prerender` then running `docker run image-id`, I get a process running `yarn prerender:work`
- When building an image with `docker build . --target prerender` then running `docker run image-id help`, I get yarn's help message

Dantemss commented on concourse/prerender-workers/script.bash:
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash -eux
```
I'll take it out and make a new PR

Dantemss committed:
Removed concourse/prerender-workers, to be re-added on the next PR

Dantemss committed:
Reverted verify-production-urls

Dantemss commented on concourse/verify-production-urls/script.bash:
```diff
@@ -6,14 +6,29 @@ production_url=https://openstax.org
 work_dir=$(pwd)
 failed=0
 
-release_id=$(curl -s "${production_url}/rex/environment.json" | jq .release_id -r)
```
I separated this into another PR and I'll continue to refactor it to move it after the uploads and to look at the buckets only. But for the purposes of this PR, it's been removed, so done.

Dantemss commented on script/prerender-workers/cfn.yml:
```diff
@@ -0,0 +1,209 @@
+AWSTemplateFormatVersion: 2010-09-09
```
I pinged him, waiting for an answer.

Dantemss commented on script/prerender-workers/manage.ts:
```diff
@@ -0,0 +1,396 @@
+const BUCKET_NAME = process.env.BUCKET_NAME || 'sandbox-unified-web-primary';
+const BUCKET_REGION = process.env.BUCKET_REGION || 'us-east-1';
+const PUBLIC_URL = process.env.PUBLIC_URL || `/rex/releases/${RELEASE_ID}`;
+const RELEASE_ID_SUFFIX = RELEASE_ID.split('/', 2)[1];
```
Yeah the ECR repo doesn't like the slashes

Dantemss commented on script/prerender-workers/manage.ts:
```diff
@@ -0,0 +1,396 @@
+      },
+      {
+        ParameterKey: 'ReleaseIdSuffix',
+        ParameterValue: RELEASE_ID_SUFFIX,
```
The worker uploads using BUCKET_REGION, BUCKET_NAME and PUBLIC_URL, so I guess that knowledge is included in the PUBLIC_URL.

Dantemss committed:
Removed unused Server export

Dantemss committed:
Added a comment explaining the sqsHeartbeat function

Dantemss committed:
Converted the task switch statement to a mapping of task functions
Replaced function factories with the use of function.bind()

Dantemss committed:
Replaced some more undefined checks with assertDefined

Dantemss committed:
Combined prerender-workers code with old prerender code without modifying old code behavior

Dantemss committed:
Combined the sitemap functions some more

Dantemss committed:
Combined the old and new page rendering code as much as possible

Dantemss commented on script/prerender-workers/contentPages.tsx:
```diff
@@ -0,0 +1,254 @@
+import FontCollector from '../../src/helpers/FontCollector';
+import { readAssetFile } from './fileUtils';
+
+export type BookMatch = {
```
I combined the prerender-workers into the old prerender folder, hopefully without modifying any of the old behavior.

Dantemss commented on script/prerender-workers/thread.ts:
```diff
@@ -0,0 +1,146 @@
+          const archivePage = await getArchivePage(page);
+          return getSitemapItemOptions(archivePage, matchPathname(page));
+        });
+        await renderSitemap(payload.slug, items);
```
Moved them into separate functions and changed to a map of task types to functions instead of a switch statement.

Dantemss committed:
Removed duplicated comment

TomWoodward commented on script/prerender/fileUtils.ts:
```diff
@@ -42,3 +47,34 @@ export const createDiskCache = <K extends string, V>(prefix: string): Cache<K, V
+async function writeS3File(contentType: string, filepath: string, contents: string) {
+  let basePath = assertDefined(process.env.PUBLIC_URL, 'PUBLIC_URL environment variable not set');
+  if (basePath[0] === '/') { basePath = basePath.slice(1); }
+  const key = `${basePath}${filepath}`;
```
the presence of this logic makes the function more than just writing files to s3, there aren't a lot of rexy files outside of the release directories but there are some, it would be nice to separate this logic into like... `writeS3File` and `prefixReleasePath` or something

Dantemss replied:
The function has been split into these 2.

TomWoodward commented on script/prerender/fileUtils.ts:
```diff
@@ -42,3 +47,34 @@ export const createDiskCache = <K extends string, V>(prefix: string): Cache<K, V
+  }));
+}
+
+export const writeS3HtmlFile = writeS3File.bind(null, 'text/html');
```
i'd have done this with function scope, with a semi curried function, rather than bind and messing around with the whole `this` context

i think that is just personal preference though i don't know if there is necessarily anything wrong with doing it this way

Dantemss replied:
MDN docs claim that bind(null) doesn't change `this` but I guess they lied:
`what()` is a function that logs `this`.

```
> obj.what = what
[Function: what]
> obj.what()
{ what: [Function: what] }
undefined
> obj.what = what.bind(null)
[Function: bound what]
> obj.what()
<ref *1> Object [global] {
...
```

I'll rewrite this not to use bind()...

Dantemss replied:
The have been rewritten not to use `bind()`.

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+  queues up each page in the queue to be prerendered, and deletes the stack at the end
+*/
+
+import './setup';
```
if this file isn't rendering any react code it probably doesn't need this setup thing

Dantemss replied:
Just removing it leads to an error, but I figured it's actually contentPages that's loading the React stuff so I moved this import there and it works.

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+const BUCKET_NAME = process.env.BUCKET_NAME || 'sandbox-unified-web-primary';
+const BUCKET_REGION = process.env.BUCKET_REGION || 'us-east-1';
+const PUBLIC_URL = process.env.PUBLIC_URL || `/rex/releases/${RELEASE_ID}`;
+const RELEASE_ID_SUFFIX = RELEASE_ID.split('/', 2)[1];
```
this doesn't really have any value when split out, i think it would make more sense to string replace the blocked characters rather than just using the last bit

Dantemss replied:
Done

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+let numBooks = 0;
+let numPages = 0;
+
+async function renderManifest() {
```
this is duplicated from `script/prerender/prerender.tsx`

Dantemss replied:
Split `renderManifest()` into its own file

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+      },
+      {
+        ParameterKey: 'ValidUntil',
+        ParameterValue: `${timeoutDate.toISOString().slice(0, -5)}Z`,
```
what are you slicing off the end here? the zone? why?

Dantemss replied:
Have to slice off the milliseconds since AWS only accepts seconds

TomWoodward replied:
we have a date uitll library with a [format](https://date-fns.org/v2.28.0/docs/format) function that you could use to construct the right thing instead of mangling the ISO string

Dantemss replied:
It seems date-fns 1.x does not support printing ISO strings if your local time is not GMT.
However, I can update it to 2.x and add the date-fns-tz library.

Dantemss replied:
I added date-fns-tz to convert the Date to UTC. I had to update date-fns because date-fns-tz fails to import with date-fns 1.x

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+  console.log(`[${book.title}] Queuing ${numBookPages} book pages in batches of 10`);
+
+  for (let pageIndex = 0; pageIndex < pages.length; pageIndex += 10) {
+    const pageBatch = pages.slice(pageIndex, pageIndex + 10);
```
you can use `'lodash/chunk'`

Dantemss replied:
Done

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+    const sendMessageBatchResult = await sqsClient.send(new SendMessageBatchCommand({
+      Entries: pageBatch.map((page, batchIndex) => ({
+        Id: batchIndex.toString(),
+        MessageBody: JSON.stringify({ payload: { page }, type: 'page' }),
```
why the wrapping page key? is there something wrong with the payload just being the `{state, params}` object?

Dantemss replied:
I did it for consistency with the sitemap type which needs them but you are right, it is not necessary so I'll remove it.

Dantemss replied:
Done

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+  // console.log(`[${book.title}] Queuing sitemap`);
+
+  await sqsClient.send(new SendMessageCommand({
+    MessageBody: JSON.stringify({ payload: { pages, slug: book.slug }, type: 'sitemap' }),
```
you don't have any typing for these queue bodies, it might be helpful to create some aliases for encoding them and some guards that match the type on read so that you can get strong typing

Dantemss replied:
I did already add the guards on read, but I'll add the types as well.

Dantemss replied:
I added types when sending messages. You may want to check the receiving side too though. I have some types there, but it mostly relies on runtime checks.

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+async function manage() {
+  console.log('Preloading route components');
+
+  await Loadable.preloadAll();
```
this is for the rendering, its not necessary here

Dantemss replied:
Removed

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+
+  await Loadable.preloadAll();
+
+  archiveLoader = createArchiveLoader(REACT_APP_ARCHIVE_URL, {
```
you should just define these above, there is no point in letting them be undefined until now

Dantemss replied:
I thought they needed the preload, but if that is not the case yeah I'll move them up.

TomWoodward replied:
no `Loadable.preloadAll();` is necessary for the react components in each route, thats how we avoid bundling all the route components at once

Dantemss replied:
Moved this code up to the spot where they are first defined

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+    stack.Outputs, '[CFN] [DescribeStacks] Unexpected response: missing stack Outputs key'
+  );
+
+  let workQueueUrl: string | undefined;
```
i get that its less efficient to loop over the outputs twice but since there are only like two outputs i feel like it would be fine to do a `find` here and it would be much more readable

Dantemss replied:
Done, replaced with 2 calls to `find()`

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+
+  console.log(`All ${numPages} page prerendering jobs and all ${numBooks} sitemap jobs queued`);
+
+  // console.log('Queuing sitemap index job');
```
commented log?

Dantemss replied:
Removed

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+  console.log('Waiting 1 minute for the work queue attributes to stabilize');
+
+  // First wait 1 minute after sending the last message for the queue attributes to stabilize
+  // This is required according to SQS docs
```
link to docs? some of this comment is redundant with the log above and can be removed

Dantemss replied:
Reorganized the comments and added links

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+  console.log('Waiting for the work queue to be empty');
+
+  // Now check that all the NumberOfMessages attributes are 0
+  const getQueueAttributesCommand = new GetQueueAttributesCommand({
```
this manage function is pretty big, i feel like you could extract this whole waiting for queue to be empty thing into a helper

Dantemss replied:
Split it into smaller pieces and as a bonus got rid of all non-const global variables in this file.

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+  // Ensure that the dead letter queue is also empty
+  // Since we are the only consumer of the dead letter queue, we use long polling to check
+  // that it is empty rather than waiting another minute and checking the queue attributes
+  const receiveDLQMessageResult = await sqsClient.send(new ReceiveMessageCommand({
```
this removes the messages from the queue right? wouldn't that prevent us from being able to redrive them?

Dantemss replied:
Normally this would make the messages invisible, but the DLQ has the VisibilityTimeout set to 0 which may or may not mean messages can never be successfully deleted from it.
However, the DLQ is deleted when the workers stack is deleted at the end of the build, even if it failed, which happens right after the manager detects any messages in the DLQ.
If you're thinking about wanting to retry failed builds (not from the beginning), then we'd maybe have to split the DLQ into a separate stack.

TomWoodward replied:
it may become convenient to have this script try to redrive the dlq once or twice before exiting if the build becomes flaky, not necessarily right now but it would be nice if we didn't have to redo this part of the logic if we decide to do that

Dantemss replied:
I think that is handled in the normal queue. If the worker doesn't delete the message it'll be retried by another worker when the visibility timeout expires. The DLQ only comes into play when the message has already failed some (configurable) number of times. To retry more, we can just increase that number instead.

TomWoodward commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+createWorkersStack().then(() => manage().catch(async(e) => {
+  // catch for manage() only
+  // Do not wait for the stack deletion to complete since we can't handle a delete failure anyway
+  await deleteWorkersStack();
```
this should be in a finally so it doesn't need to be called in two places

Dantemss replied:
Moved to a try-finally with await instead of then().finally() since that's probably more readable.

TomWoodward reviewed (changes requested):
i only made it through the manager code

Dantemss committed:
Use more threads for bigger instances

staxly[bot] committed:
Merge branch 'master' into prerender_workers

Dantemss committed:
Replaced CodeVersion and ReleaseIdSuffix parameters with ReleaseId and SanitizedReleaseId

Dantemss commented on script/prerender/cfn.yml:
```diff
@@ -0,0 +1,209 @@
+    ConstraintDescription: Must be a valid AWS region name
+
+  CodeVersion:
+    Description: Code version to use, e.g. "v4"
```
I got rid of the CodeVersion, ReleaseIdSuffix and instead added ReleaseId and SanitizedReleaseId.

Dantemss commented on script/prerender/cfn.yml:
```diff
@@ -0,0 +1,209 @@
+      QueueName: !Sub rex-${CodeVersion}-${ReleaseIdSuffix}-prerender-dlq
+      ReceiveMessageWaitTimeSeconds: 20
+      VisibilityTimeout: 0
+      # RedriveAllowPolicy: # leads to circular dependency errors with the worker queue
```
This would limit which queues can use this DeadLetterQueue. I don't see a way to set it on creation though, due to the circular dependency issue, only edit it in later.

Dantemss commented on script/prerender/cfn.yml:
```diff
@@ -0,0 +1,209 @@
+      KmsMasterKeyId: alias/aws/sqs
+      QueueName: !Sub rex-${CodeVersion}-${ReleaseIdSuffix}-prerender-work-queue
+      ReceiveMessageWaitTimeSeconds: 20
+      RedriveAllowPolicy: '{"redrivePermission": "denyAll"}'
```
Can't use the WorkQueue as another queue's dead letter queue.

Dantemss commented on script/prerender/cfn.yml:
```diff
@@ -0,0 +1,209 @@
+    DependsOn: Cluster # workaround for Cluster deletion to work properly
+    Properties:
+      KmsMasterKeyId: alias/aws/sqs
+      QueueName: !Sub rex-${CodeVersion}-${ReleaseIdSuffix}-prerender-work-queue
```
Sure, I guess. Just need to also randomize the stack name and the image tag in the ECR.

Dantemss replied:
I think I'll leave the name and just add a random suffix so we can still tell what each of the resources belongs to

Dantemss replied:
I guess the image doesn't need the random suffix, I can just allow it to be pushed again. In theory the code should be the same, just the internal docker ids may be different.

Dantemss commented on script/prerender/contentPages.tsx:
```diff
@@ -31,12 +31,39 @@ export const stats = {
+}
+
+/* Note: PageMatch can be made more restrictive if single-instance prerender code is removed:
+export type PageMatch = {
```
I used to have this type but had to take it out because of the single-instance prerender code.
I'll remove the comment.

Dantemss commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+    switch (outputKey) {
+      case 'WorkQueueUrl':
+        workQueueUrl = outputValue;
+        resolveWorkQueuePromise(workQueueUrl);
```
0 seconds since the workers take some time to boot even after the stack is deployed and the manager can start queuing up some messages before they are ready. :) Fine, I'll rewrite this.

Dantemss replied:
Done, it now waits for the stack creation at the beginning and gets the queue urls before it starts loading any of the books.

Dantemss committed:
Updated some comments

Dantemss committed:
Added a random BUILD_ID suffix, used to keep stack and resource names unique

Dantemss committed:
Changed the Application tags to Rex Prerender

Dantemss commented on script/prerender/cfn.yml:
```diff
@@ -0,0 +1,209 @@
+              - Key: Project
+                Value: Unified
+              - Key: Application
+                Value: Rex
```
Changed to "Rex Prerender"

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

Dantemss committed:
Moved content route stuff (sitemap etc) to a separate file
Rewrote some of the types using existing generic types

Dantemss commented on script/prerender/contentPages.tsx:
```diff
@@ -31,12 +31,39 @@ export const stats = {
 };
 
-export async function prepareContentPage(
+export type BookMatch = {
```
Done, changed to use `Match<Route<>>`

Dantemss commented on script/prerender/contentPages.tsx:
```diff
@@ -31,12 +31,39 @@ export const stats = {
+};
+export type SerializedBookMatch = Omit<BookMatch, 'route'>;
+
+export function deserializeBook(book: SerializedBookMatch) {
```
I did this at first but then it turned out to not be necessary. I could just delete `deserializeBook()`, since the sitemapIndex stuff did not need the `route: content`.

Dantemss commented on script/prerender/contentPages.tsx:
```diff
@@ -31,12 +31,39 @@ export const stats = {
+  route: typeof content,
+  state: { bookUid: string, bookVersion: string },
+};
+export type SerializedBookMatch = Omit<BookMatch, 'route'>;
```
Moved these types and getArchivePage/Book to prerender/contentRoutes

Dantemss committed:
Removed unnecessary assertDefineds in getArchivePage() and getArchiveBook()

Dantemss commented on script/prerender/contentPages.tsx:
```diff
@@ -118,41 +145,77 @@ export const minuteCounter = () => {
+
+  return assertDefined(
+    await services.archiveLoader.book(bookUid, bookVersion).load(),
+    'book wasn\'t cached, it should have been'
```
I removed them and typescript did not complain so I guess they were not necessary.

Dantemss committed:
Changed most contentPages functions to work with SerializedPageMatch instead of PageMatch
renderAndSavePage() now calls deserializePageMatch() to add { route: content } to its match argument

Dantemss commented on script/prerender/contentPages.tsx:
```diff
@@ -171,15 +234,20 @@ export const prepareBooks = async(
 };
 
-export type Pages = Array<{code: number, route: Match<typeof content>}>;
+// Note: Could return Array<SerializedPageMatch> if single-instance prerender code is removed
```
Removed the "code" argument and changed most of these functions to use SerializedPageMatch, only adding the content route when needed.

TomWoodward commented:
i looked at the comments but most of them aren't tagged as outdated so i think you still have some changes to make? you can press the request re-review when you're ready for me to look again

Dantemss committed:
Removed commented-out RedriveAllowPolicy

Dantemss committed:
Decomposed writeS3File function and replaced bind() calls

Dantemss committed:
Moved import './setup' to contentPages

Dantemss committed:
Split renderManifest() into its own file

Dantemss committed:
Format the ISO date using date-fns-tz instead of toISOString()

Dantemss committed:
Use lodash/fp/chunk instead of a manual loop to chunk pages

Dantemss committed:
Unnest page task and sitemapIndex task payloads

Dantemss committed:
More typechecks for message payloads

Dantemss committed:
Removed Loadable.preloadAll() from manage.ts and moved loader initialization to the beginning of the file

Dantemss committed:
Removed commented-out log messages

Dantemss committed:
Improved comments about checking that the SQS queue is empty

Dantemss committed:
Rewrote most of manage.ts to make the logic easier to follow

Dantemss commented on script/prerender/manage.ts:
```diff
@@ -0,0 +1,397 @@
+  );
+
+  // Wait for the workers stack to be ready
+  await waitUntilStackCreateComplete(
```
I couldn't move this into createWorkersStack() because I want a failure on the create command to not trigger a stack delete command, but a failure on waitUntilStackCreateComplete() to trigger a stack delete command.

However, I extracted this and the part that gets queue urls into a separate function and rewrote a lot of this file to hopefully make everything clearer.

Dantemss committed:
Rewrote findOutputValue() for clarity

Dantemss committed:
manage() pieces now pass some arguments around; got rid of all non-const globals

Dantemss committed:
Fixed undefined variable error

Dantemss committed:
Take BuildId from the stack name rather than passing it in as a parameter

Dantemss committed:
Pass the PrerenderImageTag to the workers stack rather than a SanitizedReleaseId

Dantemss committed:
Re-added CodeVersion parameter, required for the workers to run

Dantemss committed:
Increased workers stack create/delete timeout

Dantemss committed:
Don't set execArgv for worker_threads
Add a global hook to handle all unhandledRejections and exit
Split some of thread.ts into smaller functions

Dantemss committed:
Removed max-old-space-size and added vm.max_map_count instead

Dantemss committed:
Added tests for assertObject()

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

Dantemss committed:
Removed remaining commented-out console.log()

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

TomWoodward commented on script/prerender/contentPages.tsx:
```diff
@@ -118,47 +119,63 @@ export const minuteCounter = () => {
 
-  const {app, styles, state, url} = await prepareApp(services, route, code);
-  console.info(`rendering ${url}`); // tslint:disable-line:no-console
+  await savePage(url, html);
```
do you want to type this on line 130 to return a promise?

Dantemss replied:
Ah good idea

TomWoodward commented on script/prerender/fileUtils.ts:
```diff
@@ -42,3 +47,46 @@ export const createDiskCache = <K extends string, V>(prefix: string): Cache<K, V
+
+    console.log('Initializing S3 client');
+    s3Client = new S3Client({ credentials, region: process.env.BUCKET_REGION });
+  }
```
there is some danger of doing this twice concurrently, you could also do it like
```
const getS3Client = once(async () => {
    console.log('Fetching container credentials');
    const credentials = await fromContainerMetadata();

    console.log('Initializing S3 client');
    return new S3Client({ credentials, region: process.env.BUCKET_REGION });
})

async function writeS3File(key: string, contents: string, contentType: string) {
  const s3Client = await getS3Client();
}


```

this way even if its called concurrently both usages will get the same promise

Dantemss replied:
Ah interesting... ok I'll add this

TomWoodward commented on script/prerender/sitemap.ts:
```diff
@@ -7,25 +7,54 @@ import max from 'lodash/fp/max';
+
+  const filePath = sitemapPath(slug);
+
+  await saveFile(filePath, bookSitemap.toString());
```
do you want to type this as returning a promise?

Dantemss replied:
writeAssetFile() is not async but I suppose I can wrap it

TomWoodward commented on src/app/utils/assertions.ts:
```diff
@@ -27,3 +27,15 @@ export const assertString = <X>(x: X, message: string): string => {
+    throw new Error(message);
+  }
+
+  return x as Exclude<X, number | string | boolean | bigint | symbol | null | undefined>;
```
i have a different version of this https://github.com/openstax/ancillary-service/blob/main/utils/src/guards.ts#L6 i'm wondering if one or the other type conversion is better?

Dantemss replied:
I think yours is more strict, since it only allows non-class non-array non-function objects as far as I can tell. I use mine on instances of some class so I need the more lax type.

Dantemss committed:
Make savePage return Promise<unknown> and make savePageAsset async

Dantemss committed:
Use lodash's once function to initialize the S3Client

Dantemss committed:
Make saveFile return Promise<unknown> and make renderSitemap* functions use an async version of writeAssetFile

Dantemss committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

staxly[bot] committed:
Merge branch 'master' into prerender_workers

Dantemss committed:
Integrate prerender workers with Concourse (#1408)

* Readded concourse/prerender-workers

* Renamed prerender-workers/* to prerender/manage.*

* Renamed prerender to prerender:local and prerender:manage to prerender:fleet

* Renamed prerender target to release and made it the default
Removed entrypoint and cmd from release target

* Use a prerender-mode parameter to switch prerendering modes
No longer need to run yarn install and yarn build on the rex-web image and cache node_modules

* Removed -prerender suffix from Rex image tag

* Fixed lint errors

* Downcased ci target; Build release target from utils instead of CI

* Re-added removed calls to yarn install and yarn build

* Allow execute on concourse/prerender/script.bash

* Added AWS secrets and BUCKET_NAME to prerender task
Read IMAGE_TAG from build-config

* Re-added -prerender- to the prerender stack name

* Give the large Docker image another try

* Run prerender from the image, not from the repo

* Also set the correct paths for update-content

* Build Docker image with build-configs as build args
Use PUBLIC_URL during yarn build and store it in the image

* Might as well store all build configs in the image

* Try the other Docker env variable format to see if it handles quotes better

* Create unquoted config file to use with oci-build-task

* Increased the prerender stack creation timeout

* Increased the workers stack creation/deletion timeout to 5 minutes

* Revert "update content (#1468)"

This reverts commit 31c735063670ffafbd2732eee06565408ad82aa4.

* Pass the document to wrapSolutions

* Fixed the redirect for Fisica Universitaria

* Handle unicode URIs in update-redirects-data

* Also call decodeURI in prepareRedirects

* Add a maximum number of SQS heartbeats
Make failed messages immediately visible to other workers

* Reorganized and DRYed up code that changes ReceiptHandle visibility timeout

Co-authored-by: staxly[bot] <35789409+staxly[bot]@users.noreply.github.com>
Co-authored-by: Thomas Woodward <tom.g.woodward@gmail.com>
